tests/suites/os/tests/secureboot/index.js

```diff
@@ -6,50 +6,113 @@ module.exports = {
 		{
 			title: 'check secure boot',
 			run: async function(test) {
+				const securebootEfiVarPath = '/sys/firmware/efi/efivars/SecureBoot-*';
+
 				const securebootSupported = await this.worker.executeCommandInHostOS(
-					['ls', '/sys/firmware/efi/efivars/SecureBoot-*', '&>/dev/null',
+					['ls', securebootEfiVarPath, '&>/dev/null',
 						'&&', 'echo', 'true',
 						'||', 'echo', 'false'
 					],
 					this.link,
 				).then(output => { return output === 'true'});
 
+				// https://wiki.archlinux.org/title/Unified_Extensible_Firmware_Interface/Secure_Boot#Checking_Secure_Boot_status
+				const securebootEnabled = await this.worker.executeCommandInHostOS(
+						[
+							'od', '--address-radix=n', '--format=u1', securebootEfiVarPath,
+								'|', 'tr', '-s', "' '",
+								'|', 'cut', "-d' '", '-f6'
+						],
+						this.link,
+				).then(output => output === '1' );
+
+				// Used after tampering with verified boot files
+				const resetWorker = async () => {
+					await this.worker.off();
+					await this.worker.flash(this.os.image.path);
+					await this.worker.on();
+
+					await this.worker.addSSHKey(this.sshKeyPath);
+				}
+
+				const waitForSerialOutput = async(pattern, slice=0, retries=60, delay=1000) => this.utils.waitUntil(
+					async () => this.worker.fetchSerial().then(
+						serialLogs => serialLogs.split('\n')
+																		.slice(slice)
+																		.join('\n')
+																		.match(pattern)
+					), false, retries, delay
+				);
+
```

Comment:
@alexgg This is an area I think we could do better in Leviathan, defining features commonly used by tests, fully developing them as well thought-out features and exposing them as optional modules.

In this particular case, this function is useful, but I don't think its usage and semantics are well-defined enough to become an interface. In particular, it's hamstrung by the `worker.fetchSerial()` method pulling the complete serial log from the `core` container. Because of the entire serial log being returned for each call, I added the `slice` parameter, which allows for only examining a slice of the text, such as the last N lines, which is useful when we want to know if the most recent serial output matches what we're looking for. However, we're still transferring that full log over HTTP, and I predict some usages will suffer as a result.

I feel the ideal interface here would be event driven, and test writers would be able to receive lines of text as they're logged. The bare minimum Leviathan would need to do then is expose a socket that serial logs can be read from, but the fragmented architecture with multiple containers communicating over HTTP makes ideas like this more difficult to implement than necessary.

Reply:
Right, thanks for the context. @rcooke-warwick @vipulgupta2048 having serial functions available to all suites is probably core Leviathan work that would be nice to implement, as I said above a test for serial console availability in development mode is still lacking and will need this, as well as this test.